### ISSUE TITLE
Update MongoHealthCheck for new Mongo driver version

### ISSUE THREAD

**Issue by sleberknight** [kiwiproject/dropwizard-service-utilities#233]
`MongoHealthCheck` uses the Mongo `com.mongodb.DB` which exists in the Mongo 3.x driver but not in the 4.x and higher versions.

It needs to be replaced, which changes the public API of this class, specifically the constructor.

**Comment by sleberknight** [comment]
See the following migration guide for the Mongo driver:

https://www.mongodb.com/docs/drivers/java/sync/current/legacy/#api-changes

**Comment by sleberknight** [comment]
After some initial investigation, I found that the replacement for `DB` is `MongoDatabase`, which doesn't have some of the features the health check uses. Specifically, it does not have an equivalent to the `getMongo()` method, which we used to obtain the Mongo host/port information:

```java
var host = db.getMongo().getAddress().toString();
```

We also used the `getStats()` method which returned a `CommandResult` object. Neither this method nor this class exist anymore in the new Mongo driver. The `getStats()` method did this:

```java
public CommandResult getStats() {
  BsonDocument commandDocument = new BsonDocument("dbStats", new BsonInt32(1)).append("scale", new BsonInt32(1));
  return executeCommand(commandDocument);
}
```

We used it to call the `ok()` method which returns a `boolean` to determine if healthy or not. We also used it to obtain the `"storageSize"` and `"dataSize"` properties which we included in the details of the health check results. And finally, if there was an error, we called the `getErrorMessage()` method to obtain an error message.

If we want to retain the same or similar functionality, we need to determine how to replicate this functionality. We can of course copy the code to create the `BsonDocument` for the `dbStats` command. And the `MongoDatabase` class does contain a bunch of overloaded `runCommand` methods some of which return a `Document` and some which return a generic type that you specify in an argument.

Basically, we just need to figure out how to execute the [dbStats](https://www.mongodb.com/docs/manual/reference/command/dbStats/#mongodb-dbcommand-dbcmd.dbStats) command as described in the Mongo documentation. With this new and "improved" `MongoDatabase`, it looks like it will be _more_ work and custom code, not less, unless I can find that maybe they just moved all this functionality somewhere else. I am not counting on that, however.

**Comment by sleberknight** [comment]
However, after reading more about the [dbStats](https://www.mongodb.com/docs/manual/reference/command/dbStats/#mongodb-dbcommand-dbcmd.dbStats) command, I'm not sure if we _should_ replicate all the functionality that we had been using. One concern is this:

> The time required to run the command depends on the total size of the database. Because the command must touch all data files, the command may take several seconds to run.

We don't want to be doing this _every_ time this health check is called. If we really want this functionality, then we should probably cache the storage and data size information for some (possibly configurable) amount of time. While the cache is not expired, we simply return the last known values. No idea what a good default would be for the cache expiration, e.g. maybe 10-15 minutes?

Another concern is this regarding "unexpected" shutdowns:

> After an unclean shutdown of a [mongod](https://www.mongodb.com/docs/manual/reference/program/mongod/#mongodb-binary-bin.mongod) using the [Wired Tiger](https://www.mongodb.com/docs/manual/core/wiredtiger/) storage engine, count and size statistics reported by [dbStats](https://www.mongodb.com/docs/manual/reference/method/db.stats/#mongodb-dbcommand-dbcmd.dbStats) may be inaccurate.

So, the stats themselves might not even be accurate.

For the health check to be meaningful, we need to find _something_ to do to validate that the Mongo database is available. We'll need to do more research into alternative commands we could use. A good starting point might be [Diagnostic Commands](https://www.mongodb.com/docs/manual/reference/command/nav-diagnostic/). For example, a few commands we might consider include `hostInfo` and `ping` and `serverStatus`.

@chrisrohr and/or @terezivy - any thoughts on all this?

**Comment by chrisrohr** [comment]
If all we want is the status of the server, then the `ping` command should be good.  I agree, we don't want to add process intensive healthchecks.

**Comment by sleberknight** [comment]
Here's a reference to running commands in MongoDB:

https://www.mongodb.com/docs/manual/tutorial/use-database-commands/

In the [Command Responses](https://www.mongodb.com/docs/manual/tutorial/use-database-commands/#command-responses) section, it only mentions that there is always an "ok" property defined as:

> A number that indicates whether the command has succeeded (1) or failed (0).

Here's the result of running a successful `ping` command:

```json
{
  "operationTime": Timestamp(7088351293026074, 1),
  "ok": 1.0,
}
```

In the previous Mongo 3.x driver there is a wrapper class, `CommandResult`, that encapsulates some of the logic to determine success or failure. A few of the key methods are `ok()` which returns a `boolean`, and `getErrorMessage()` which returns a nullable `String`.

`ok()` is defined as:

```java
    /**
     * Gets the "ok" field, which is whether this command executed correctly or not.
     *
     * @return true if the command executed without error.
     */
    public boolean ok() {
        Object okValue = get("ok");
        if (okValue instanceof Boolean) {
            return (Boolean) okValue;
        } else if (okValue instanceof Number) {
            return ((Number) okValue).intValue() == 1;
        } else {
            return false;
        }
    }
```

I'm not really sure why they are checking if the value was an instance of `Boolean` if the result is always either zero or 1, unless maybe earlier Mongo versions might return different values for `ok`.

And `getErrorMessage()` is defined as:

```java
    /**
     * Gets the error message associated with a failed command.
     *
     * @return The error message or null
     */
    @Nullable
    public String getErrorMessage() {
        Object foo = get("errmsg");
        if (foo == null) {
            return null;
        }
        return foo.toString();
    }
```

By executing various commands in actual Mongo databases, I've observed the following:

* `ok` is a `Double` with value 0.0 or 1.0
* `errmsg` is a `String`
* when `errmsg` exists, there are two more properties: `code` and `codeName` which describe the error code and name, respectively.

Example of a bad command:

```javascript
db.runCommand( { crash: 1 })
```

with result:

```json
{
  "operationTime": Timestamp(7088351293026074, 1),
  "ok": 0.0,
  "errmsg": "no such command: 'crash'",
  "code": 59,
  "codeName": "CommandNotFound"
}
```

So, our health check can just use the `ping` command and replicate the logic in the `CommandResult` class. It can also include the additional `code` and `codeName` details if the ping fails.
